Modules/Module_Processing/Module_ASL/xASL_wrp_ResampleASL.m

```diff
@@ -254,84 +253,90 @@ function xASL_wrp_ResampleASL(x)
-			
-            % MultiPLD PWI after averaging
-			PWI = zeros(size(ASL_im,1), size(ASL_im,2), size(ASL_im,3), max(indexNew)); % preallocate PWI
-            for nPLD = 1:max(indexNew)
-                PWI(:, :, :, nPLD) = xASL_stat_MeanNan(ASL_im(:, :, :, indexNew == nPLD), 4); % Averaged PWI4D 
+			% indexAverage_PLD_LabDur lists for each original position to where it should be averaged
+			[~, ~, iUnique_PLD_LabDur] = unique(unique([initialPLD_PWI4D, LabDuration_PWI4D], 'stable', 'rows'), 'stable', 'rows');
+
+            % MultiPLD-multiLabDur PWI3D after averaging
+			PWI3D = zeros(size(PWI4D,1), size(PWI4D,2), size(PWI4D,3), max(iUnique_PLD_LabDur)); % preallocate PWI
+            for iPLD_LabDur = 1:max(iUnique_PLD_LabDur)
+                PWI3D(:, :, :, iPLD_LabDur) = xASL_stat_MeanNan(PWI4D(:, :, :, iUnique_PLD_LabDur == iPLD_LabDur), 4); % Averaged PWI4D 
             end
-            
-            % Save PWI
-            fprintf('%s\n', PathPWI4D{iSpace});
-            xASL_io_SaveNifti(PathASL{iSpace}, PathPWI4D{iSpace}, PWI, 32, false);
-            
-            PWIsingle = xASL_stat_MeanNan(ASL_im, 4); % create single PWI for further steps in ASL module
-            
-            % Save single PWI
-            fprintf('%s\n', PathPWI{iSpace});
-            xASL_io_SaveNifti(PathASL{iSpace}, PathPWI{iSpace}, PWIsingle, 32, false);
-            
         else
-            PWI = xASL_stat_MeanNan(ASL_im, 4); % singlePLD PWI
-            % Save PWI
-            fprintf('%s\n', PathPWI{iSpace});
-            xASL_io_SaveNifti(PathASL{iSpace}, PathPWI{iSpace}, PWI, 32, false);
-        end            
+            PWI4D = ASL_im;
+        end
+        
+        PWI3D = xASL_stat_MeanNan(PWI4D, 4);
     end
+    
+    
+    % =====================================================================
+    % D) Save subtracted to disk
+
+    % Save PWI (single volume)
+    fprintf('%s\n', PathPWI{iSpace});
+    xASL_io_SaveNifti(PathASL4D{iSpace}, PathPWI{iSpace}, PWI, 32, false);
+
+    % Save PWI3D (averaged volume for each PLD-labdur combination)
+    fprintf('%s\n', PathPWI{iSpace});
+    xASL_io_SaveNifti(PathASL4D{iSpace}, PathPWI3D{iSpace}, PWI3D, 32, false);
+    
+    % Save PWI4D (subtracted only, not yet averaged)
+    fprintf('%s\n', PathPWI4D{iSpace});
+    xASL_io_SaveNifti(PathASL4D{iSpace}, PathPWI4D{iSpace}, PWI4D, 32, false);    
 end
 
 
 %% ------------------------------------------------------------------------------------------
 % 9. Save PWI NIfTI & time-series-related maps (SD, SNR)
 MaskIM = xASL_io_Nifti2Im(fullfile(x.D.MapsSPMmodifiedDir, 'rgrey.nii'));
 MaskIM = MaskIM>(0.7*max(MaskIM(:)));
-PWI = xASL_stat_MeanNan(ASL_im, 4); % repeat this here, in case it is skipped above for single volumes (e.g. GE 3D spiral)
+PWI4D = xASL_stat_MeanNan(ASL_im, 4); % repeat this here, in case it is skipped above for single volumes (e.g. GE 3D spiral)

```

Comment:
I'm confused - why PWI4D when this is an average? Isn't this supposed to be PWI3D?

Reply:
I didn't fix this yet, because here we need to either:
1) run this only for the first TE and last PLD/labdur (then we would perhaps create a temporary PWI4D for the first TE/last PLD only)
2) run this for each TE/PLD/labdur combination

Now started fixing this in 1e23fac0b573d5e119411f1bf640c80ed901d8a3

Reply:
Please close a question like this if it is clear to you (and in the issue list for fixing)

Reply:
This is now removed. So can be resolved.